go/protocol/chat1/extras.go

```diff
@@ -915,6 +916,27 @@ func (d *SignEncryptedData) AsSealed() SealedData {
 	}
 }
 
+func (d EncryptedData) CanonicalHash() []byte {
+	return d.AsSealed().CanonicalHash()
+}
+
+func (d SignEncryptedData) CanonicalHash() []byte {
+	return d.AsSealed().CanonicalHash()
+}
+
+func (d SealedData) CanonicalHash() []byte {
+	state := sha256.New()
+	// Everything here is either fixed length (8 bytes for the version number)
+	// or length-prefixed (the nonce and the encrypted payload). That
+	// guarantees that the serialization cannot collide with a different input.
+	binary.Write(state, binary.LittleEndian, uint64(d.V))
+	binary.Write(state, binary.LittleEndian, uint64(len(d.N)))
+	binary.Write(state, binary.LittleEndian, d.N)
+	binary.Write(state, binary.LittleEndian, uint64(len(d.E)))
+	binary.Write(state, binary.LittleEndian, d.E)
+	return state.Sum(nil)
```

Comment:
Would it be good enough to just serialize a msgpack array? Also what do prev ptrs do?

Reply:
scratch that, I should be using `makeHeaderHash`. though I might want to make it stricter about nonce lengths